2899376 refactor queue consumption

fgm opened the pull request:
https://www.drupal.org/node/2899376

fgm committed:
Add first attempt at Travis integration.

fgm committed:
Tests passing locally.

fgm committed:
Added Travis-CI badge for 'travis' branch. Remember to update it for 8.x-1.x when merging.

fgm committed:
Changes from phcorp's review.

fgm committed:
Issue #2904149 by fgm: improve coding standards compliance.

fgm committed:
Issue #2904149 by fgm: include Travis coverage on Scrutinizer.

phcorp committed:
fix `QueueBase::getQueue()` unused `$options` parameter

phcorp committed:
Fix comments, assertions in test.

phcorp committed:
implement `QueueBase::shutdown()` method

fgm committed:
Add hook_requirements() support and rabbitmq properties page.

fgm committed:
Cleanup queue deletions.

fgm committed:
Avoid autoloading issues during install at per #2667588.

fgm committed:
Support separate error conditions for unconfigured credentials and queue service.

fgm committed:
Merge branch '2899376-refactor' into travis

fgm committed:
Issue #2899376 by fgm: refactor consumer to a service.

fgm committed:
Issue #2899376: actually support timeouts in basic_consume().

phcorp committed:
shutdown queue

fgm committed:
Issue #2899376 by fgm,phcorp: Do not request QOS in a consume loop.

- Removed echo, typos
- Added comments.

fgm committed:
Merge pull request #16 from phcorp/shutdown-queue

Shutdown queue

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/ed49faac-ecef-4ce3-8d49-a8855bba9968) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/ed49faac-ecef-4ce3-8d49-a8855bba9968) completed: **17 new issues, 34 updated code elements**

Saphyel commented:
@FGM there is any BC changes?

fgm commented:
There shouldn't be, 

* unless you count the fact that asking for a timeout previously didn't actually trigger one in most situations, while it now actually works. I know at least one site which actually used this non-operational setting
* or the fact that tests are no longer broken (...)
* Also, there's a new rabbitmq-timeout parameter on the drush rqwk command.
* There's a new DecoderAwareWorkerInterface which QueueWorker plugins can implement, in order to have a custom job item decoder implementation, to fix the recurring issue of having a single encoding format in this module.

So I rather think of this as compatible new features rather than BC issues.

fgm commented:
@Saphyel Anything you'd like changed in order to have it merged on d.o. ?

Saphyel commented:
I'm testing it at the moment on my projects, so if I see that still behaves as before I'll merge it.. bear in mind that you did a lot of changes but I'm reviewing it

fgm commented:
OK, not a problem: it's good to see it being tested !

Saphyel commented on composer.json:
```diff
@@ -9,6 +9,10 @@
     "source": "http://cgit.drupalcode.org/rabbitmq"
   },
   "require": {
+    "ext-pcntl": "*",
```
@FGM why do you require this ?
You don't use it anywhere in the code that extension as far I see ... ?

fgm replied:
It is the (core bundled) extension containing the `pcntl_alarm`/`pcntl_signal` functions which make the timeout mechanism actually work regardless of the driver behavior. They are used in `Consumer::consume()`.

Saphyel replied:
@FGM Why do you want to add that extension? if they stop to release/maintain it in php 7.3 would be an issue, and people that can't install/enable extensions this could be a problem

fgm replied:
This is need to make the timeout mechanism actually work regardless of the driver behavior. Also, this extension is bundled by default in PHP, what makes you think it could become deprecated in a future version beyond PHP 7.2 ?

Saphyel replied:
@FGM sorry I didn't mean that I know what is gonna happen in the future ... I mean completely the opposite.

Maybe is better if you split this PR into 2? one with the refactor and fixes and another one with the implementation of that extension and I can make a proper release with that change and people can discuss if that is ok or not...

I have several sites in maintenance that they don't have that extension enabled, and probably I'm not the only one with that issue

fgm replied:
Well, I don't have time to do this right now, preparing for DC Vienna and with the site going live next week. But I wouldn't object if someone did that split, of course. Keeping in mind that the current timeout system (included in another d.o. issue) just doesn't work in the general case; we've had lots of unexpected problems in production with the current version, and this fixes the issue by going one level lower. 

I'm very surprised some sites don't have that extension since it is bundled with PHP by default, so sites don't need to load it as they have it already. But maybe a better way would be to mark it as recommended instead of required, and wrap the pcntl-dependent calls in an availability check ?

fgm replied:
Also, I just noticed that the underlying driver add extra signals with that extension (see https://github.com/php-amqplib/php-amqplib/blob/master/README.md), so we could extend the handling further in a followup.

Saphyel replied:
yep, I think would be better talk after the DC, enjoy the last DC europe!!

fgm replied:
Hopfully not the last. But maybe the last under the auspices of the US D.A.

fgm committed:
Made PCNTL extension optional.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/7779df84-f2f1-47f0-a4a0-e005e1d048a2) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/d4bfb2f6-85eb-4807-be26-0039ebab4257) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/d4bfb2f6-85eb-4807-be26-0039ebab4257) completed: **16 new issues, 36 updated code elements**

fgm commented:
@Saphyel as you can see, pcntl is now optional.

fgm committed:
Various minor errors and coding standards issues.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/dd78b28a-b068-4ba3-8c4f-7d84bcaa78ba) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/dd78b28a-b068-4ba3-8c4f-7d84bcaa78ba) completed: **6 new issues, 36 updated code elements**

fgm committed:
Implement a Queue API-based consumer in addition to the AMQP one.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/355cff3e-80c4-415f-a6e0-667fbd646e2e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/355cff3e-80c4-415f-a6e0-667fbd646e2e) completed: **9 new issues, 37 updated code elements**

fgm committed:
Mark lack of RabbitMQ settings as warning instead of error.

- Avoid causing installation fails in situations where build settings differ from run settings.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/236d5f8a-dac3-4f9c-bc27-860ffe9bbc1b) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/FGM/rabbitmq/inspections/236d5f8a-dac3-4f9c-bc27-860ffe9bbc1b) completed: **9 new issues, 37 updated code elements**
